File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.java


import java.util.Arrays;
import java.util.Scanner;
import java.util.Set;
import java.util.TreeSet;

public class Main {

	public static void main(String[] args) {

		Scanner sc = new Scanner(System.in);

		int n = sc.nextInt();

		int[] arr = new int[n + 1];
		int[] dp = new int[n + 1];

		for (int i = 1; i < n + 1; i++) {
			arr[i] = sc.nextInt();
		}

		for (int i = 1; i < n + 1; i++) {
			for (int j = 0; j < i; j++) {
				if (arr[i] > arr[j]) {
					dp[i] = Math.max(dp[j] + 1, dp[i]);
				}
			}
		}

//		System.out.println(Arrays.toString(arr));
//		System.out.println(Arrays.toString(dp));

		int ans = 0;
		for (int ele : dp) {
			ans = Math.max(ans, ele);
		}

		System.out.println(ans);

	}

}

File: 백준/Silver/11053. 가장 긴 증가하는 부분 수열/가장 긴 증가하는 부분 수열.py
'''
내가 못풀어서 답 보고 풀었음

[해결방안]
    1. dp는 1로 채운다. 왜냐하면 나부터 시작할 수 있으니까!
    2. 나보다 앞에 있는 애들 중에 내가 걔보다 크면
    -> 나는 걔 더하기 1과 전에 왔던 나중에 큰 값이된다!
'''

n = int(input())
arr = list(map(int,input().split()))
dp = [1]*n

for i in range(1,n):
    for j in range(i):
        if(arr[i]>arr[j]):
            dp[i] = max(dp[i], dp[j]+1)

# print(dp)
print(max(dp))
